feat(aws/netflix): add diff view to ASG details

icfantv opened the pull request:
* add a new item to server group info section that displays a modal
containing all the code and/or JAR changes from the pipeline that
deployed to the selected ASG.
* updated the deploy stage pipeline view to combine commit and JAR
changes to utilize same dialog for viewing the diffs.

![screen shot 2017-04-05 at 5 15 11 pm](https://cloud.githubusercontent.com/assets/370054/24732095/7595b4d8-1a23-11e7-82f5-82df2f392420.png)
![screen shot 2017-04-05 at 3 25 51 pm](https://cloud.githubusercontent.com/assets/370054/24732056/2bfded4a-1a23-11e7-815f-ac83fba9b0de.png)
![screen shot 2017-04-05 at 4 13 20 pm](https://cloud.githubusercontent.com/assets/370054/24732058/31e7a8c2-1a23-11e7-8a8e-2661993e315f.png)

icfantv commented on app/scripts/modules/amazon/serverGroup/details/serverGroupDetails.aws.controller.js:
```diff
@@ -418,5 +425,46 @@ module.exports = angular.module('spinnaker.serverGroup.details.aws.controller',
+          };
+          this.commits = [];
+
+          const deployStages = executionDetails.stages.filter(stage => stage.type === 'deploy');
```
The thought here is that pipelines can have multiple deploy stages so the code below here essentially aggregates the data from the multiple stages.  If that logic is not right, PLMK.

ajordens replied:
I can't remember ... did we include the stageId in the `spinnaker:metadata`? If so that might help narrow down to the specific deploy stage?

anotherchrisberry replied:
Yeah, you should have `stageId` - we use it to create the source link. Not sure if there are cases where there would be multiple deploy stages deploying different artifacts, but it would be safer to pick the deploy stage from the tag.

icfantv replied:
Sweet.  I'll look for the `stageId` and go from there.

icfantv replied:
Using `stageId` works and GREATLY reduces the amount of code needed to process the execution details.  Good tip you guys.  Thanks.

ajordens replied:
👍

icfantv commented:
@anotherchrisberry @jrsquared @christopherthielen PTAL.

The `Changes` item will not show if there's no execution metadata or the execution ID from the metadata doesn't return stuff.  This means that for deploy tasks not executed through pipelines (I _think_ this is possible, in theory), this feature will not be present.

ajordens commented:
Visually I think this is pretty slick, an improvement over what it was originally.

anotherchrisberry commented on app/scripts/modules/amazon/serverGroup/details/serverGroupDetails.aws.controller.js:
```diff
@@ -418,5 +425,46 @@ module.exports = angular.module('spinnaker.serverGroup.details.aws.controller',
       });
     };
+
+    this.lookForPipelineCommitDiffs = (executionId) => {
```
It would be nice to move this logic into the `viewChangesLink` component instead of pushing more stuff into this controller, which is already > 400 LOC.

anotherchrisberry commented on app/scripts/modules/core/diffs/commitHistory.component.ts:
```diff
@@ -0,0 +1,34 @@
+    commits: '<'
+  };
+  public template = `
+  <div class="step-section-details" ng-if="$ctrl.commits && $ctrl.commits.length">
```
Is the `step-section-details` class doing anything here?

icfantv replied:
It was giving padding and margin, but I think is an artifact of playing around with a bunch of different layouts/L&Fs.  I'll remove and see if goes all `GridBagLayout` on me.

anotherchrisberry commented on app/scripts/modules/core/diffs/commitHistory.component.ts:
```diff
@@ -0,0 +1,34 @@
+      <tr ng-repeat="commit in $ctrl.commits">
+        <td>{{commit.timestamp | date:'MM/dd'}}&nbsp;&nbsp;</td>
+        <td><a target="_blank" href="{{commit.commitUrl}}">{{commit.displayId}}</a>&nbsp;&nbsp;</td>
+        <td>{{commit.message | limitTo: 50}}&nbsp;&nbsp;</td>
```
Are the `&nbsp;`s doing anything here?

icfantv replied:
They were residual's from the original template and I suspect were for spacing when used in the pipeline execution details view.  I'm betting they can be safely removed.  Will check.

anotherchrisberry commented on app/scripts/modules/core/diffs/jarDiff.component.ts:
```diff
@@ -0,0 +1,82 @@
+import {IComponentController, module} from 'angular';
+
+import {IComponentOptions} from 'angular';
```
can combine these import statements

icfantv replied:
Whoops.  Good catch.

christopherthielen replied:
option-apple-O, btw

icfantv replied:
@christopherthielen that reformatted my hard drive making me lose everything.  so thanks for that.

anotherchrisberry commented on app/scripts/modules/core/diffs/jarDiff.component.ts:
```diff
@@ -0,0 +1,82 @@
+        <td>{{jar.displayDiff}}&nbsp;&nbsp;</td>
+      </tr>
+    </table>
+  </div>
```
same questions about class and spaces

anotherchrisberry commented on app/scripts/modules/core/diffs/viewChangesLink.component.ts:
```diff
@@ -0,0 +1,72 @@
+  constructor(private $uibModal: IModalService) {}
+
+  public $onInit(): void {
+    this.linkText = this.linkText || 'View Changes';
```
If you set `linkText` to `'View Changes'` in the declaration, do you still need the `$onInit` method? Or does that not work to default a value?

christopherthielen replied:
that approach works nicely for defaults

icfantv replied:
Short answer?  No.

Longer answer:  this is an artifact of trying to customize the L&F of a reusable component in several different areas.  The text `View Changes` (title case) in the SERVER GROUP INFORMATION section looks bad and doesn't match the text we use for the pipeline link which is all lowercase.

Did I understand/answer your question?

anotherchrisberry replied:
Sorry, to clarify: if you change line 32 to:
```
public linkText = 'View Changes';
```
do you still need the `$onInit` method, i.e. does Angular replace the value if it's passed in to the component?

icfantv replied:
Ahhhh, that's a very good question.  I don't know, but suspect you may be right.

anotherchrisberry commented on app/scripts/modules/core/pipeline/config/stages/deploy/deployExecutionDetails.html:
```diff
@@ -101,60 +101,9 @@
+  <div class="step-section-details" ng-if="detailsSection === 'changes'">
+    <view-changes-link commits="stage.context.commits"
+                       jar-diffs="stage.context.jarDiffs"
+                       name-item="stage.context"></view-changes-link>
```
Does the `stage.context` include a `name` field on a deploy stage?

icfantv replied:
Yes, it's the name of the deploy stage, which, in hindsight, may not actually be useful.  Maybe the Pipeline name would be better?  I didn't dig for it on the object so I don't now offhand if it's there.

anotherchrisberry replied:
Huh, I don't see `name` in the `context` of the stage itself. I only checked a few - are you sure there's a `name` field in the deploy stage context?

icfantv replied:
It's been a few hours since I committed this; lemme double check again.  I know I saw a `name` property _somewhere_.

anotherchrisberry commented:
What's going on with the duplicate JAR changes sections in the screenshot?

icfantv commented:
@anotherchrisberry the extra duplicate JAR section was an easy way to get more height and double check the `max-height` CSS property and overflow scrolling.  But now that I think about it.  I'm wondering if we should make the headers sticky then because in views like the one you gave me for nccp-server, there was a lot of stuff and so better UX might dictate sticky headers.  Thoughts?

anotherchrisberry commented:
Sticky headers wouldn't hurt - they can be finicky, but I can work with you tomorrow if they don't "just work" in that modal.

icfantv commented:
Ok, I'll have a go and see what happens.  WCGW?

icfantv commented:
Updated addressing all concerns.  Passing `stage` instead of `stage.context` for the pipeline view from which to retrieve the `name` property.  Sticky headers added for scrolling.  Pipeline execution logic moved to the View Changes Link component.

anotherchrisberry commented on app/scripts/modules/amazon/serverGroup/details/serverGroupDetails.aws.controller.js:
```diff
@@ -52,6 +52,7 @@ module.exports = angular.module('spinnaker.serverGroup.details.aws.controller',
                                                      serverGroupWarningMessageService, overrideRegistry) {
 
     this.state = {
+      changesAvailable: false,
```
doesn't look like this is ever getting set (can probably remove)

icfantv replied:
Oh yea, that's residual from the link business logic.  Will 86 it.

anotherchrisberry commented on app/scripts/modules/amazon/serverGroup/details/serverGroupDetails.html:
```diff
@@ -89,6 +89,11 @@ <h3 select-on-dbl-click>
+        <dd ng-if="ctrl.state.changesAvailable">
+          <view-changes-link commits="ctrl.commits" jar-diffs="ctrl.jarDiffs" link-text="view changes"
+                             name-item="ctrl.serverGroup"></view-changes-link>
+        </dd>
```
probably want to remove the `<dt>` and `<dd>` and move the `<view-changes-link>` into the `<render-if-feature>` body above

anotherchrisberry commented:
Couple of minor things but LGTM otherwise - feel free to merge once the `changesAvailable` flag bits are addressed

icfantv commented:
Updated and waiting for CI build to finish.
